src/Control/Effect/Interpret.hs

```diff
@@ -23,27 +77,27 @@ import Control.Monad.Trans.Class
 --   At time of writing, a simple passthrough use of 'runInterpret' to handle a 'State' effect is about five times slower than using 'StateC' directly.
 --
 --   prop> run (runInterpret (\ op -> case op of { Get k -> k a ; Put _ k -> k }) get) === a
-runInterpret :: (forall x . eff m x -> m x) -> InterpretC eff m a -> m a
-runInterpret handler = runReader (Handler handler) . runInterpretC
-
-newtype InterpretC eff m a = InterpretC { runInterpretC :: ReaderC (Handler eff m) m a }
-  deriving (Alternative, Applicative, Functor, Monad, MonadFail, MonadFix, MonadIO, MonadPlus)
+runInterpret
+  :: forall eff m a.
+     ( HFunctor eff, Monad m )
+  => ( forall x. eff m x -> m x )
+  -> ( forall s. Reifies s ( Handler eff m ) => InterpretC s eff m a )
+  -> m a
+runInterpret f m =
+  reify ( Handler handler ) ( go m )
 
-instance MonadTrans (InterpretC eff) where
-  lift = InterpretC . lift
-  {-# INLINE lift #-}
+  where
 
-newtype Handler eff m = Handler (forall x . eff m x -> m x)
+    handler :: forall s x. eff ( InterpretC s eff m ) x -> InterpretC s eff m x
```

Comment:
This signature can be removed; in fact, we can change the line above to `reify (Handler (InterpretC . f . handleCoercible)) (go m)` if we want (tho I don’t mind either way).

Reply:
That would be ambiguous, as `go` wouldn't know which `s` to chose. It does unfortunately need to be spelled out like this.